src/pluto_eris/fields/fp2.rs

```diff
@@ -0,0 +1,771 @@
+        self.c0.ct_eq(&other.c0) & self.c1.ct_eq(&other.c1)
+    }
+}
+
+impl Default for Fp2 {
+    #[inline]
+    fn default() -> Self {
+        Self::ZERO
+    }
+}
+
+impl From<Fp2> for [u8; 112] {
+    fn from(value: Fp2) -> [u8; 112] {
+        value.to_bytes()
+    }
+}
+
+impl<'a> From<&'a Fp2> for [u8; 112] {
+    fn from(value: &'a Fp2) -> [u8; 112] {
+        value.to_bytes()
+    }
+}
+
+impl Neg for Fp2 {
+    type Output = Fp2;
+
+    #[inline]
+    fn neg(self) -> Fp2 {
+        -&self
+    }
+}
+
+impl<'a> Neg for &'a Fp2 {
+    type Output = Fp2;
+
+    #[inline]
+    fn neg(self) -> Fp2 {
+        self.neg()
+    }
+}
+
+impl<'a, 'b> Sub<&'b Fp2> for &'a Fp2 {
+    type Output = Fp2;
+
+    #[inline]
+    fn sub(self, rhs: &'b Fp2) -> Fp2 {
+        self.sub(rhs)
+    }
+}
+
+impl<'a, 'b> Add<&'b Fp2> for &'a Fp2 {
+    type Output = Fp2;
+
+    #[inline]
+    fn add(self, rhs: &'b Fp2) -> Fp2 {
+        self.add(rhs)
+    }
+}
+
+impl<'a, 'b> Mul<&'b Fp2> for &'a Fp2 {
+    type Output = Fp2;
+
+    #[inline]
+    fn mul(self, rhs: &'b Fp2) -> Fp2 {
+        self.mul(rhs)
+    }
+}
+
+use crate::{
+    impl_add_binop_specify_output, impl_binops_additive, impl_binops_additive_specify_output,
+    impl_binops_multiplicative, impl_binops_multiplicative_mixed, impl_sub_binop_specify_output,
+    impl_sum_prod,
+};
+impl_binops_additive!(Fp2, Fp2);
+impl_binops_multiplicative!(Fp2, Fp2);
+impl_sum_prod!(Fp2);
+
+const SIZE: usize = 112;
+const COORD_SIZE: usize = 56;
+
+impl Fp2 {
+    #[inline]
+    pub const fn zero() -> Fp2 {
+        Fp2 {
+            c0: Fp::zero(),
+            c1: Fp::zero(),
+        }
+    }
+
+    #[inline]
+    pub const fn one() -> Fp2 {
+        Fp2 {
+            c0: Fp::one(),
+            c1: Fp::zero(),
+        }
+    }
+
+    pub const fn new(c0: Fp, c1: Fp) -> Self {
+        Fp2 { c0, c1 }
+    }
+
+    pub const fn size() -> usize {
+        SIZE
+    }
+
+    /// Attempts to convert a little-endian byte representation of
+    /// a scalar into a `Fp`, failing if the input is not canonical.
+    pub fn from_bytes(bytes: &[u8; SIZE]) -> CtOption<Fp2> {
+        let c0 = Fp::from_bytes(bytes[0..COORD_SIZE].try_into().unwrap());
+        let c1 = Fp::from_bytes(bytes[COORD_SIZE..SIZE].try_into().unwrap());
+        CtOption::new(
+            Fp2 {
+                c0: c0.unwrap(),
+                c1: c1.unwrap(),
+            },
+            c0.is_some() & c1.is_some(),
+        )
+    }
+
+    /// Converts an element of `Fp` into a byte representation in
+    /// little-endian byte order.
+    pub fn to_bytes(self) -> [u8; SIZE] {
+        let mut res = [0u8; SIZE];
+        let c0_bytes = self.c0.to_bytes();
+        let c1_bytes = self.c1.to_bytes();
+        res[0..COORD_SIZE].copy_from_slice(&c0_bytes[..]);
+        res[COORD_SIZE..SIZE].copy_from_slice(&c1_bytes[..]);
+        res
+    }
+
+    // TODO: This is a naive method using 4 multiplications
```

Comment:
When the non-residue used is `i` with `i^2 = -1` the algorithm can be improved to use 3 muls instead of 4. However, I think it's not possible to make such optimization in this case.
I'll look into it and remove the comment / apply the optimization if possible.

Reply:
Can you link the issue created for this and mark resolved?